### PR TITLE
Find if two line segments intersect

### PR DESCRIPTION
**Pull request by jaebradley**

### CODE REVIEW

File: geometry/intermediate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                                                                                            line_segment_1.end)
+
+    return is_one_end_point_collinear and \
+           (could_line_segment_1_contain_line_segment_2_start or could_line_segment_1_contain_line_segment_2_end or
+            could_line_segment_2_contain_line_segment_1_start or could_line_segment_2_contain_line_segment_1_end)
+
+
+def calculate_point_triplet_orientation(coordinate_1, coordinate_2, coordinate_3):
+    """
+    Returns a value that represents if connecting coordinate_1 to coordinate_2 to coordinate_3 occurs in a
```
Comment (jaebradley) [inline geometry/intermediate.py]: This definition could be better I think this value really represents the `-sine` of the number of degrees that the three coordinates create

File: geometry/intermediate.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, start, end):
+        self.start = start
+        self.end = end
+
+
+def do_line_segments_intersect(line_segment_1, line_segment_2):
+    """
+    Checks to see if two line segments have at least one point that they share
+    """
+    line_segment_2_start_orientation_relative_to_line_segment_1 = calculate_point_triplet_orientation(line_segment_1.start, line_segment_1.end, line_segment_2.start)
```
Comment (visola) [inline geometry/intermediate.py]: Couldn't think of a longer name uh?

Comment (visola) [reply geometry/intermediate.py]: What do you think about calling them `triplet_{segment_number}_{segment_number}_start|end}`. Something like: `triplet_1_2_start`, `triplet_1_2_end`, etc. ?

File: tests/geometry/test_intermediate.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    Unit Test for identifying if two line segments intersect
+    """
+    def test_proper_line_segments_intersect(self):
+        """
+        Tests that two proper line segments which intersect, returns True
+        """
+        line_segment_1_start = Coordinate(-1, 0)
+        line_segment_1_end = Coordinate(1, 0)
+        line_segment_1 = LineSegment(line_segment_1_start, line_segment_1_end)
```
Comment (visola) [inline tests/geometry/test_intermediate.py]: I would just re-write all these to be:
```python
line_segment_1 = LineSegment(Coordinate(-1,0), Coordinate(1, 0))
```
My reasoning is that you'll have less code lines, and less code to read means less stuff to keep in your head.
This way it makes it more explicit. But it also makes assignments that aren't necessary.
It's all a matter of preference.

File: geometry/intermediate.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    line_segment_2_end_orientation_relative_to_line_segment_1 = calculate_point_triplet_orientation(line_segment_1.start, line_segment_1.end, line_segment_2.end)
+    line_segment_1_start_orientation_relative_to_line_segment_2 = calculate_point_triplet_orientation(line_segment_2.start, line_segment_2.end, line_segment_1.start)
+    line_segment_1_end_orientation_relative_to_line_segment_2 = calculate_point_triplet_orientation(line_segment_2.start, line_segment_2.end, line_segment_1.end)
+
+    does_line_segment_2_have_one_end_clockwise_to_line_segment_1 = line_segment_2_start_orientation_relative_to_line_segment_1 * line_segment_2_end_orientation_relative_to_line_segment_1 < 0
+    does_line_segment_1_have_one_end_clockwise_to_line_segment_2 = line_segment_1_start_orientation_relative_to_line_segment_2 * line_segment_1_end_orientation_relative_to_line_segment_2 < 0
+
+    if does_line_segment_2_have_one_end_clockwise_to_line_segment_1 and does_line_segment_1_have_one_end_clockwise_to_line_segment_2:
```
Comment (visola) [inline geometry/intermediate.py]: I bet that if I draw this I would understand, but it's pretty complicated to see, even with the extra long variable names.

**Review by visola — APPROVED** [review]
Too much math for me to validate this by head. I believe in your test cases and since your code coverage says you're at 100%, sounds like it's being thoroughly tested.
But as I said in a comment, this math is more complicated than I can see/reason about.

File: geometry/intermediate.py [hunk 5]
```diff
@@ -12,39 +12,59 @@ def __init__(self, start, end):
 
 def do_line_segments_intersect(line_segment_1, line_segment_2):
     """
-    Checks to see if two line segments have at least one point that they share
+    Checks to see if two line segments have at least one point that they share.
+
+    Let's call the ends of line_segment_1, A and B, and the ends of line_segment_2, C and D.
```
Comment (visola) [inline geometry/intermediate.py]: This is great! Now I see it!

File: geometry/intermediate.py [hunk 6]
```diff
@@ -12,39 +12,59 @@ def __init__(self, start, end):
...
+               could_segment_1_contain_segment_2_start or could_segment_1_contain_segment_2_end or
+               could_segment_2_contain_segment_1_start or could_segment_2_contain_segment_1_end
+           )
 
 
 def calculate_point_triplet_orientation(coordinate_1, coordinate_2, coordinate_3):
     """
-    Returns a value that represents if connecting coordinate_1 to coordinate_2 to coordinate_3 occurs in a
-    clockwise (value < 0), counterclockwise (value > 0), or collinear (value = 0) fashion.
+    Returns a value that represents the negative sine of the angle created by connecting connecting coordinate_1 to
```
Comment (visola) [inline geometry/intermediate.py]: typo: `connecting connecting`, duplicate.

File: geometry/intermediate.py [hunk 7]
```diff
@@ -12,39 +12,59 @@ def __init__(self, start, end):
...
+    coordinate_2 to coordinate_3.
+
+    If this connection occurs in a clockwise fashion the value will be negative, counterclockwise will be positive,
+    and collinear will be 0.
+
+    This is equivalent to taking the cross-product of the vectors created by the first and second coordinates and
```
Comment (visola) [inline geometry/intermediate.py:66]: Now you're just showing off. 😄

**Review by visola — APPROVED** [review]
Much much better! Now I understand it!

**Pull request merged**